optuna/visualization/_pareto_front.py

```diff
@@ -288,3 +408,65 @@ def _make_hovertext(trial: FrozenTrial) -> str:
         indent=2,
     )
     return text.replace("\n", "<br>")
+
+
+def _split_trials_by_constraints(
+    trials: List[FrozenTrial],
+    n_best_trials: int,
+    constraints_func: Callable[[FrozenTrial], Sequence[float]],
+) -> Tuple[List[FrozenTrial], List[FrozenTrial], List[FrozenTrial]]:
+    best_feasible_trials = []
+    non_best_feasible_trials = []
+    infeasible_trials = []
+
+    for i, trial in enumerate(trials):
+        constraint_score = constraints_func(trial)
+        is_feasible = all(map(lambda x: x <= 0.0, constraint_score))
+        if is_feasible:
+            if i < n_best_trials:
+                best_feasible_trials.append(trial)
+            else:
+                non_best_feasible_trials.append(trial)
+        else:
+            infeasible_trials.append(trial)
+
+    return best_feasible_trials, non_best_feasible_trials, infeasible_trials
+
+
+def _make_marker(
+    trials: List[FrozenTrial],
+    include_dominated_trials: bool,
+    use_constraints_func: bool,
+    dominated_samples: bool = False,
+    color: Optional[str] = None,
+) -> Dict[str, Any]:
+    if dominated_samples and not include_dominated_trials:
+        assert len(trials) == 0
+
+    if not use_constraints_func:
+        if dominated_samples:
+            return {
+                "line": {"width": 0.5, "color": "Grey"},
+                "color": [t.number for t in trials],
+                "colorscale": "Blues",
+                "colorbar": {
+                    "title": "#Trials",
+                },
+            }
+        else:
+            return {
+                "line": {"width": 0.5, "color": "Grey"},
+                "color": [t.number for t in trials],
+                "colorscale": "Reds",
+                "colorbar": {
+                    "title": "#Best trials",
+                    "x": 1.1 if include_dominated_trials else 1,
+                    "xpad": 40,
+                },
+            }
+    else:
+        assert color is not None
+        return {
+            "line": {"width": 0.5, "color": "Grey"},
+            "color": color,
+        }
```

Comment:
If `use_constraints_func==True`, `include_dominated_trials` seems to be ignored. I guess users may want to extract only the best feasible. What do you think of it?

Reply:
In this case, non pareto front trials are excluded in `_get_pareto_front_2d` or `_get_pareto_front_3d` (`_make_marker` need not care about `include_dominated_trials`).
However, pareto front but infeasible trials are still included. Do you mean this?

Reply:
Thank you for your explanation. When I wrote this comment, I thought some parts of the marker style should be changed. But I found that we don't need that since we don't use `colorscale` if `constraints_func` is specified.

> However, pareto front but infeasible trials are still included. Do you mean this?

As we talked in the dev sprint, we need to remove the infeasible solutions before extracting the Pareto solutions. Let's discuss this topic in #3128.